### PR TITLE
msg/async: avoid put message within write_lock

### PR DESCRIPTION
**Pull request by yuyuyu101**
message deconstruct under busy env isn't a very short period, because cpu may
stuck into tcmalloc library to do gc.

from a rough bench(3k iops), reduce send_message latency from 54us to 37.5us. in
real env(higher iops), it should be much more better.

Signed-off-by: Haomai Wang <haomai@xsky.com>

### CODE REVIEW

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -2299,15 +2299,20 @@ void AsyncConnection::handle_ack(uint64_t seq)
...
   while (!sent.empty() && sent.front()->get_seq() <= seq) {
     Message* m = sent.front();
     sent.pop_front();
+    pending.pendingpush_back(m);
     ldout(async_msgr->cct, 10) << __func__ << " got ack seq "
                                << seq << " >= " << m->get_seq() << " on "
                                << m << " " << *m << dendl;
-    m->put();
   }
+  write_lock.write_lockunlock();
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc]: s/write_lockunlock/unlock()/

File: src/msg/async/AsyncConnection.cc [hunk 2]
```diff
@@ -2299,15 +2299,20 @@ void AsyncConnection::handle_ack(uint64_t seq)
 {
   ldout(async_msgr->cct, 15) << __func__ << " got ack seq " << seq << dendl;
   // trim sent list
-  std::lock_guard<std::mutex> l(write_lock);
+  std::vector<Message*> pending;
```
Comment (liewegas) [inline src/msg/async/AsyncConnection.cc]: small_vector here?

Comment (yuyuyu101) [reply src/msg/async/AsyncConnection.cc]: boost/container/small_vector.hpp?

Comment (liewegas) [reply src/msg/async/AsyncConnection.cc]: yeah.  it'll avoid an allocation in the general case

**Comment by yuyuyu101** [comment]
I looked at small_vector, I think c-style array is enough to handle as well as boost::small_vector

**Comment by branch-predictor** [comment]
@yuyuyu101 From what I see, if the number messages to drop will exceed max_pending, any outstanding messages will remain in queue until the next ack. Won't that be an issue?

**Comment by liewegas** [comment]
@yuyuyu101 ping

**Comment by yuyuyu101** [comment]
@branch-predictor actually it's impossible that we will left message in sent queue, although it's not a problem still. each seq ack will call handle_ack to deal with sent queue, so it doesn't happen message left in the "sent" even when idle.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-04-16_09:06:16-rados-wip-kefu-testing-2018-04-16-1515-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-04-15_23:41:27-rados-wip-kefu-testing-2018-04-15-2317-distro-basic-smithi/

the failed tests are caused by #21338

**Pull request merged**